src/tools/spark-sstfile-generator/src/main/scala/com/vesoft/tools/SstFileOutputFormat.scala

```diff
@@ -0,0 +1,233 @@
+
+  /**
+    * need local file system only, for rocksdb sstFileWriter can't write to remote file system(only can back up to HDFS) for now
+    */
+  private val localFileSystem = FileSystem.get(new Configuration(false))
+
+  /**
+    * hdfs file system, for create destination sst file dir
+    */
+  private val hdfsFileSystem = FileSystem.get(configuration)
+
+  if (!hdfsFileSystem.getScheme.equalsIgnoreCase("hdfs")) {
+    throw new IllegalStateException("File system is not hdfs")
+  }
+
+  /**
+    * which dir in hdfs to put sst files
+    */
+  private val hdfsParentDir = configuration.get(SparkSstFileGenerator.SSF_OUTPUT_HDFS_DIR_CONF_KEY)
+
+  log.debug(s"SstRecordWriter read hdfs dir:${hdfsParentDir}")
+
+  // all RocksObject should be closed
+  private var env: EnvOptions = _
+  private var options: Options = _
+
+  override def write(key: GraphPartitionIdAndKeyValueEncoded, value: PropertyValueAndTypeWritable): Unit = {
+    var sstFileWriter: SstFileWriter = null
+
+    //cache per partition per vertex/edge type's sstfile writer
+    val sstWriterOptional = sstFilesMap.get((value.vertexOrEdgeEnum, key.partitionId))
+    if (sstWriterOptional.isEmpty) {
+      /**
+        * https://github.com/facebook/rocksdb/wiki/Creating-and-Ingesting-SST-files
+        *
+        * Please note that:
+        *
+        *1. Options passed to SstFileWriter will be used to figure out the table type, compression options, etc that will be used to create the SST file.
+        *2. The Comparator that is passed to the SstFileWriter must be exactly the same as the Comparator used in the DB that this file will be ingested into.
+        *3. Rows must be inserted in a strictly increasing order.
+        */
+      //TODO: how to make sure the options used here to be consistent with the server's
+      // What's the fastest way to load data into RocksDB: https://rocksdb.org.cn/doc/RocksDB-FAQ.html
+      env = new EnvOptions
+      // TODO: prove setWritableFileMaxBufferSize is working?
+      options = new Options().setCreateIfMissing(true).setCreateMissingColumnFamilies(true).setWritableFileMaxBufferSize(1024 * 100).setMaxBackgroundFlushes(5).prepareForBulkLoad()
+      sstFileWriter = new SstFileWriter(env, options)
+
+      // TODO: rolling to another file when file size > some THRESHOLD, or some other criteria
+      // Each partition can generated multiple sst files, among which keys will be ordered, and keys could overlap between different sst files.
+      // All these sst files will be  `hdfs -copyFromLocal` to the same HDFS dir(and consumed by subsequent nebula `IMPORT` command), so we need different suffixes to distinguish between them.
+      val hdfsSubDirectory = s"${File.separator}${key.partitionId}${File.separator}"
+
+      val localDir = s"${localSstFileOutput}${hdfsSubDirectory}"
+      val sstFileName = s"${value.vertexOrEdgeEnum}-${key.`type`}-${DatatypeConverter.printHexBinary(key.valueEncoded.getBytes)}.sst".toLowerCase
```

Comment:
We should design the SSTFile's format, it should included the partition number for `download` & `ingest`.

Reply:
@darionyaphet  
FYI:
It will generate multiple sst file for a single partition, for it will be TOO LARGE to generate just one sst file per partition when the partition number is low.

Check the class level comment to see the local dir structure which hold the sst files before be `hdfs -copyFromLocal`

```
   worker_nodeX
   |
   |-sstFileOutput
   |
      |--1 (*THIS IS PARTITION NUMBER*)
      |  |
      |  |——vertex-${FIRST_KEY}.sst
      |  |--edge-${FIRST_KEY}.sst
      |
      |--2
         |
         |——vertex-${FIRST_KEY}.sst
         |--edge-${FIRST_KEY}.sst
```

For the reason mentioned above, i name different sst file with vertex/edge-${FIRST_KEY}.sst to tell them apart.

Reply:
Could we store the hdfs files as partition number,  such as:
part-1
     | ----  file1
     | ----  file2
     ....
part-2
part-3
....

Reply:
If there is just 1 worker, that's ok. But there are multiple sst files/per partition/per worker, every worker node have a `file1` under dir `part-1`, When `hdfs -copyFromLocal`, same sst file under the same path will override each other.

Think about those :

```
worker_node1
   |---part-1
            | ----  file1 (This will be put under hdfs dir:  hdfs://host:port/some_root_dir/part-1/file1)
            | ----  file2
     ....
worker_node2
   |---part-1
            | ----  file1(This will be also put under hdfs dir:  hdfs://host:port/some_root_dir/part-1/file1, they share the same hdfs directory)
            | ----  file2
     ....
```

Reply:
Could we use worker-node1-file1 as the filename?

Reply:
Have talked with @dangleptr , doc the missing part:

Actually this is the local dir structure of _EACH_ worker node:
```
${LOCAL_ROOT} (local dir will be stripped when `hdfs -copyFromLocal`)
   |---1 (this is PARTITION number)
            | ----  vertex-${FIRST_KEY_IN_THIS_FILE}.sst
            | ---- vertex-${FIRST_KEY_IN_THIS_FILE}.sst
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst
   |---2
            | ----  vertex-${FIRST_KEY_IN_THIS_FILE}.sst
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst
     ....
```

Sst file name convention is `{TYPE}-${FIRST_KEY_IN_THIS_FILE}.sst`, where type=vertex OR edge,  `FIRST_KEY_IN_THIS_FILE` is the first key put in it. This composition will make the sst file name unique between all worker nodes.

After `hdfs -copyFromLocal` ,the final hdfs dir layout is:
```
${HDFS_ROOT}
   |---1 (this is PARTITION number, will hold all sst file from every single worker node with the same PARTITION number)
            | ----  vertex-${FIRST_KEY_IN_THIS_FILE}.sst(may be from worker node#1)
            | ---- vertex-${FIRST_KEY_IN_THIS_FILE}.sst(may be from worker node#2)
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst(may be from worker node#1)
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst(may be from worker node#2)
   |---2 (same as above)
            | ----  vertex-${FIRST_KEY_IN_THIS_FILE}.sst
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst
            | ---- edge-${FIRST_KEY_IN_THIS_FILE}.sst
     ....

```